Add CIDR and User Agent processors (#4585)
opensearch-project/documentation-website: _data-prepper/pipelines/configuration/processors/user-agent.md

---
layout: default
title: user_agent
parent: Processors
grand_parent: Pipelines
nav_order: 130
---

# user_agent

The `user_agent` processor parses any user agent (UA) string in an event then adds the parsing results to the event's write data.

## Usage

> In this example, the `user_agent` processor calls the source that contains the UA string, the `ua` field, and indicates the key where the parsed string will write, `user_agent`.

Reviewer: Is this first example of the `.yml` file illustrating the parsing of the user agent string in an event, as described in the first statement of the "# user_agent" definition above? And then does the example in JSON  illustrate the parsing of the results to the event's write data in the second statement in the "# user_agent" definition? If yes, can you somehow say that this first example performs a first step and the second example performs a second step in the # user_agent" definition? 

Or, is the .yml file example a configuration step while the JSON example fulfills both the parsing and the adding/writing process? 

I guess what I'm asking is better delineation between what the two examples are achieving.

Author: This should be resolved now.

```yaml
  processor:
    - user_agent:
        source: "ua"
        target: "user_agent"
```

> If, for example, the following event contain the `ua` field contains the following UA string:

Reviewer 2: ```suggestion
If, for example, the following event contains the `ua` field contains the following UA string:
```

> If, for example, the following event contain the `ua` field contains the following UA string:

Reviewer 2: There are words missing in this sentence, either after "contains" or "field".

> If, for example, the following event contain the `ua` field contains the following UA string:

Author: @natebower: How does this sound?
```suggestion
The following example event contains the `ua` field with the a string that details information about a user: 
```

Reviewer 2: How about that?

```json
{
  "ua":  "Mozilla/5.0 (iPhone; CPU iPhone OS 13_5_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1.1 Mobile/15E148 Safari/604.1"
}
```

> The `user_agent` processors parses the string into a format compatible with Elastic Common Schema (ECS), then adds the result to the specified target, as showing in the following example:

Reviewer: ```suggestion
The `user_agent` processor parses the string into a format compatible with Elastic Common Schema (ECS), then adds the result to the specified target, as shown in the following example:
```

> The `user_agent` processors parses the string into a format compatible with Elastic Common Schema (ECS), then adds the result to the specified target, as shown in the following example:

Reviewer 2: ```suggestion
The `user_agent` processor parses the string into a format compatible with Elastic Common Schema (ECS) and then adds the result to the specified target, as shown in the following example:
```

```json
{
  "user_agent": {
    "original": "Mozilla/5.0 (iPhone; CPU iPhone OS 13_5_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1.1 Mobile/15E148 Safari/604.1",
    "os": {
        "version": "13.5.1",
        "full": "iOS 13.5.1",
        "name": "iOS"
    },
    "name": "Mobile Safari",
    "version": "13.1.1",
    "device": {
        "name": "iPhone"
    }
  },
  "ua":  "Mozilla/5.0 (iPhone; CPU iPhone OS 13_5_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1.1 Mobile/15E148 Safari/604.1"
}
```

## Configuration options

You can use the following configuration options with the `user_agent` processor.

| Option | Required | Description |
| :--- | :--- | :--- |
| `source` | Yes | The field in the event that will be parsed. 

> | `target` | No | The field where the parsed event will write. Default is `user_agent`. 

Reviewer 2: ```suggestion
| `target` | No | The field to which the parsed event will write. Default is `user_agent`. 
```
